ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,68 @@ let host_supports_hvm ~__context host =
 let env_with_path env_vars =
   Array.of_list @@
   List.map (fun (k,v) -> Printf.sprintf "%s=%s" k v) (("PATH", String.concat ":" Forkhelpers.default_path)::env_vars)
+
+module Task  = struct (* can't place these functions in task helpers due to circular dependencies *)
+  let wait_for ~__context ~tasks =
+    let our_task = Context.get_task_id __context in
+    (** note that in this definition of `classes` we might duplicate our_task
+      * it turns out that filtering out our_task breaks _at least_ `VM.revert`, so best to leave it like this *)
+    let classes = List.map (fun x -> Printf.sprintf "task/%s" (Ref.string_of x)) (our_task::tasks) in
+    let rec process token =
+      TaskHelper.exn_if_cancelling ~__context; (* First check if _we_ have been cancelled *)
+      let statuses = List.filter_map (fun task -> try Some (Db.Task.get_status ~__context ~self:task) with _ -> None) tasks in
+      let unfinished = List.exists (fun state -> state = `pending) statuses in
+      if unfinished
+      then begin
+        let from = call_api_functions ~__context
+            (fun rpc session_id -> Client.Client.Event.from ~rpc ~session_id ~classes ~token ~timeout:30.0) in
+        debug "Using events to wait for tasks: %s" (String.concat "," classes);
+        let from = Event_types.event_from_of_rpc from in
+        process from.Event_types.token
+      end else
+        ()
+    in
+    process ""
+
+  let to_result ~__context ~of_rpc ~name ~t =
+    wait_for ~__context ~tasks:[t];
+    let fail msg = raise (Api_errors.Server_error (Api_errors.internal_error, [Ref.string_of t; Printf.sprintf "%s: %s"name msg])) in
```

Comment:
you could use `Api_errors.(Server_error (internal_error, ...))` to avoid repetition, likewise further down